lib/browser/api/power-monitor.ts

```diff
@@ -10,19 +10,35 @@ Object.setPrototypeOf(PowerMonitor.prototype, EventEmitter.prototype)
 
 const powerMonitor = createLazyInstance(createPowerMonitor, PowerMonitor, true)
 
-// On Linux we need to call blockShutdown() to subscribe to shutdown event.
 if (process.platform === 'linux') {
-  powerMonitor.on('newListener', (event:string) => {
-    if (event === 'shutdown' && powerMonitor.listenerCount('shutdown') === 0) {
-      powerMonitor.blockShutdown()
-    }
-  })
+  /* In order to delay system shutdown when e.preventDefault() is invoked
+   * on a powerMonitor 'shutdown' event, we need an org.freedesktop.login1
+   * delay lock. For more details see the "Taking Delay Locks" section of
+   * https://www.freedesktop.org/wiki/Software/systemd/inhibit/
+   *
+   * So here we watch for 'shutdown' listeners to be added or removed and
+   * set or unset our shutdown delay lock accordingly. */
+  const onceReady = (): void => {
+    powerMonitor.on('newListener', (event: string) => {
+      // whenever the listener count is incremented to one...
+      if (event === 'shutdown' && powerMonitor.listenerCount('shutdown') === 0) {
+        powerMonitor.blockShutdown()
+      }
+    })
+    powerMonitor.on('removeListener', (event: string) => {
+      // whenever the listener count is decremented to zero...
+      if (event === 'shutdown' && powerMonitor.listenerCount('shutdown') === 0) {
+        powerMonitor.unblockShutdown()
+      }
+    })
+  }
 
-  powerMonitor.on('removeListener', (event: string) => {
-    if (event === 'shutdown' && powerMonitor.listenerCount('shutdown') === 0) {
-      powerMonitor.unblockShutdown()
-    }
-  })
+  const { app } = require('electron')
+  if (app.isReady()) {
+    onceReady()
+  } else {
+    app.once('ready', onceReady)
```

Comment:
Using `app.whenReady` can make this simpler.

Reply:
> Using `app.whenReady` can make this simpler.

Yay for `app.whenReady`. So it would be the same thing as I suggested initially, ie using `app.promises.ready`. But I was not aware of `app.whenReady`, same as `app.isReady()`.

Reply:
This is a great idea. 6798825a8f1c49cb4e40dfb57fa5d42fac8cb0b2